### PR TITLE
fix: concurrent map writes error in device manager

### LINKED ISSUE #102517: kubelet panic with fatal error: concurrent map writes
/sig node
/area kubelet

#### What happened:
kubelet panic with fatal error concurrent map writes when allocate container device

* panic log
```shell
Jun  1 19:47:24 gpu-node-330 kubelet: fatal error: concurrent map writes
Jun  1 19:47:24 gpu-node-330 kubelet: goroutine 2684 [running]:
Jun  1 19:47:24 gpu-node-330 kubelet: runtime.throw(0x434c2b8, 0x15)
Jun  1 19:47:24 gpu-node-330 kubelet: /usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc00223f6c8 sp=0xc00223f698 pc=0x432eb2
Jun  1 19:47:24 gpu-node-330 kubelet: runtime.mapassign_faststr(0x3a96760, 0xc000c3b4a0, 0xc000e62fc0, 0x24, 0x7029de0)
Jun  1 19:47:24 gpu-node-330 kubelet: /usr/local/go/src/runtime/map_faststr.go:211 +0x417 fp=0xc00223f730 sp=0xc00223f6c8 pc=0x416d27
Jun  1 19:47:24 gpu-node-330 kubelet: k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).Allocate(0xc000c944e0, 0xc000903c00, 0xc002687760, 0x42, 0xc00223fb78)
Jun  1 19:47:24 gpu-node-330 kubelet: /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:365 +0x592 fp=0xc00223faa8 sp=0xc00223f730 pc=0x1b5b692
```

#### What you expected to happen:
kubelet no panic when allocate container device

#### How to reproduce it (as minimally and precisely as possible):
* Create pod, container has extended resource (like nvidia.com/gpu)
* Kubelet panic occasionally on nodes

#### Anything else we need to know?:
* system message logs
[messages.log](https://github.com/kubernetes/kubernetes/files/6581782/messages.log)

#### Environment:
- Kubernetes version (use `kubectl version`):
```shell
root@gpu-node-330:~$ kubelet --version
Kubernetes v1.18.9
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```shell
root@gpu-node-330:~$ clear
root@gpu-node-330:~$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`):
```shell
root@gpu-node-330:~$ uname -a
Linux gpu-node-330 3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
* nvidia device plugin
```shell
 $ kubectl -n kube-system get ds nvidia-device-plugin-daemonset -o wide
NAME                             DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE    CONTAINERS                 IMAGES                                                          SELECTOR
nvidia-device-plugin-daemonset   2         2         2       2            2           accelerate-type=nvidia   432d   nvidia-device-plugin-ctr   k8s-device-plugin:1.0.0-beta   name=nvidia-device-plugin-ds
```
- Network plugin and version (if this is a network-related bug):
- Others:

### LINKED ISSUE #103838: DeviceManager panic on Allocate()
#### What happened:
Pod worker panic when trying to re-allocating the device plugin resource during the node restart. 

The issue is potentially caused by [m.devicesToReuse map deleting](https://github.com/kubernetes/kubernetes/blob/1a9ae3454993b2362503db3efe1d7e48307bf3ff/pkg/kubelet/cm/devicemanager/manager.go#L376) an un-matching pod id. The map will be later used in an [assignment](https://github.com/kubernetes/kubernetes/blob/1a9ae3454993b2362503db3efe1d7e48307bf3ff/pkg/kubelet/cm/devicemanager/pod_devices.go#L158), which caused panic. ([Allocate()](https://github.com/kubernetes/kubernetes/blob/1a9ae3454993b2362503db3efe1d7e48307bf3ff/pkg/kubelet/cm/devicemanager/manager.go#L369) func by itself is not thread-safe)

Stack trace:
```
runtime.go:78] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
goroutine 1419 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x3cd4a80, 0x4c848b0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x3cd4a80, 0x4c848b0)
 /usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.podDevices.removeContainerAllocatedResources(0xc000960000, 0xc00140e240, 0x24, 0xc000c789f0, 0x9, 0x0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/pod_devices.go:109 +0x1b2
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).Allocate(0xc000995790, 0xc001475800, 0xc001302dc0, 0x42, 0xc001a21d88)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:388 +0x373
k8s.io/kubernetes/pkg/kubelet/cm/devicemanager.(*ManagerImpl).GetDeviceRunContainerOptions(0xc000995790, 0xc001475800, 0xc001302dc0, 0x0, 0x0, 0x0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/devicemanager/manager.go:881 +0x5af
k8s.io/kubernetes/pkg/kubelet/cm.(*containerManagerImpl).GetResources(0xc000430fc0, 0xc001475800, 0xc001302dc0, 0x70, 0x70, 0x42f7840)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/container_manager_linux.go:661 +0x83
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).GenerateRunContainerOptions(0xc000ec2a80, 0xc001475800, 0xc001302dc0, 0xc001d0a904, 0xb, 0xc00221d0c0, 0x1, 0x1, 0x3, 0x0, ...)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet_pods.go:455 +0x6c
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).generateContainerConfig(0xc0000ff900, 0xc001302dc0, 0xc001475800, 0x1, 0xc001d0a904, 0xb, 0xc0014aeeb0, 0x47, 0xc00221d0c0, 0x1, ...)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:247 +0xa7
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).startContainer(0xc0000ff900, 0xc00219c000, 0x40, 0xc0001985b0, 0xc00221d190, 0xc001475800, 0xc0009a1b00, 0xc000191180, 0x1, 0x1, ...)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_container.go:166 +0x5bf
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod.func1(0x458330a, 0x9, 0xc00221d190, 0xc000198500, 0x0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:816 +0x8e3
k8s.io/kubernetes/pkg/kubelet/kuberuntime.(*kubeGenericRuntimeManager).SyncPod(0xc0000ff900, 0xc001475800, 0xc0009a1b00, 0xc000191180, 0x1, 0x1, 0xc000d20dc0, 0xc0021afda0, 0x4, 0x4, ...)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kuberuntime/kuberuntime_manager.go:855 +0xc24
k8s.io/kubernetes/pkg/kubelet.(*Kubelet).syncPod(0xc000ec2a80, 0x0, 0xc001475800, 0x0, 0xc0009a1b00, 0x0, 0xc000db1440, 0xc000604660)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1719 +0x1018
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop.func1(0xc0021fdea0, 0xc00036a460, 0xc000f4fe88, 0x4d35e00, 0xc000a354c0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:175 +0x25d
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).managePodLoop(0xc00036a460, 0xc001c947e0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:184 +0x13f
k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod.func1(0xc00036a460, 0xc001c947e0)
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:222 +0x6c
created by k8s.io/kubernetes/pkg/kubelet.(*podWorkers).UpdatePod
 /workspace/louhi_ws/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/pod_workers.go:220 +0x3d3
```

#### What you expected to happen:
DeviceManager Allocate() should be able to handle such race safely. 

#### How to reproduce it (as minimally and precisely as possible):
Customer reported issue after upgrading to 1.18+. 

#### Anything else we need to know?:
This is the new behavior introduced since 1.18+ from [this commit](https://github.com/kubernetes/kubernetes/commit/cb9fdc49db8ed56aea6569e1a9ab69920ae8cc2e).

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug
/sig node

#### Which issue(s) this PR fixes:
Fixes #102517 
Fixes #103838

#### Does this PR introduce a user-facing change?

```release-note
kubelet: fix concurrent map writes error in device manager
```

### CODE REVIEW

**Comment by klueska** [comment]
Have you seen this actually cause an issue? This variable is only used in `Allocate()` and all calls to `Allocate()` are serialized.

**Comment by pacoxu** [comment]
> Have you seen this actually cause an issue? This variable is only used in `Allocate()` and all calls to `Allocate()` are serialized.

According to log https://github.com/kubernetes/kubernetes/files/6581782/messages.log, `the concururrent map failure` would be caused by this line https://github.com/kubernetes/kubernetes/blob/v1.18.9/pkg/kubelet/cm/devicemanager/manager.go#L365 

I cannot find another goroutine running to this method/function in the log. It's weried.

Can this be golang issue? Or I may ask the issue author to test with new kubelet versions.

**Comment by klueska** [comment]
The new refactoring is definitely better, but I’m still concerned about the need dir this at all. If the calls to Allocate() are not serialized, then the logic here has bigger issues than simply triggering a fault on a concurrent map access.

**Comment by pacoxu** [comment]
> The new refactoring is definitely better, but I’m still concerned about the need for this at all. If calls to Allocate() are not serialized, then the logic here has bigger issues than simply triggering a fault on a concurrent map access.

Agree. I think we should see more evidence from @tongchao199, or I need try to reproduce it with latest version and nvidia pluign. 

/hold 
for more evidence

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 1]
```diff
@@ -364,36 +364,55 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
 }
 
+func (m *ManagerImpl) getDevicesToReuse(podUID string) (map[string]sets.String, bool) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	devices, ok := m.devicesToReuse[podUID]
+	return devices, ok
+}
+
+func (m *ManagerImpl) seDevicesToReuse(podUID string, devices map[string]sets.String) {
```
Comment (adisky) [inline pkg/kubelet/cm/devicemanager/manager.go:374]: is it `setDevicesToReuse`?

Comment (pacoxu) [reply pkg/kubelet/cm/devicemanager/manager.go:374]: typoo. 😅
I will fix it.

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 2]
```diff
@@ -364,36 +364,55 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
+	defer m.mutex.Unlock()
+	delete(m.devicesToReuse, podUID)
+}
+
 // Allocate is the call that you can use to allocate a set of devices
 // from the registered device plugins.
 func (m *ManagerImpl) Allocate(pod *v1.Pod, container *v1.Container) error {
-	if _, ok := m.devicesToReuse[string(pod.UID)]; !ok {
-		m.devicesToReuse[string(pod.UID)] = make(map[string]sets.String)
+	if _, ok := m.getDevicesToReuse(string(pod.UID)); !ok {
```
Comment (chenyw1990) [inline pkg/kubelet/cm/devicemanager/manager.go:389]: I think get and set should be handled in the same function.
just like this：

devicesToReuse := m.createDevicesToReuseIfNotExist(string(pod.UID));

func (m *ManagerImpl) createDevicesToReuseIfNotExist(podUID string) map[string]sets.String {
	m.mutex.Lock()
	defer m.mutex.Unlock()
	if _, ok := m.devicesToReuse[podUID]; !ok {
            m.devicesToReuse[podUID] = make(map[string]sets.String)
        }
	return m.devicesToReuse[podUID]
}

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 3]
```diff
@@ -364,36 +364,55 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
...
+		m.seDevicesToReuse(string(pod.UID), make(map[string]sets.String))
 	}
 	// If pod entries to m.devicesToReuse other than the current pod exist, delete them.
 	for podUID := range m.devicesToReuse {
```
Comment (chenyw1990) [inline pkg/kubelet/cm/devicemanager/manager.go:393]: Should we lock this piece of code to protect it?
```	
for podUID := range m.devicesToReuse {
		if podUID != string(pod.UID) {
			m.clearDevicesToReuse(podUID)
		}
	}
```

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 2 above)
Comment (pacoxu) [reply pkg/kubelet/cm/devicemanager/manager.go:389]: Good idea. 
The set is only used for initialization.

File: pkg/kubelet/cm/devicemanager/manager.go (hunk 3 above)
Comment (pacoxu) [reply pkg/kubelet/cm/devicemanager/manager.go:377]: I refactor it.

**Comment by pacoxu** [comment]
The fix is updated. I tested several times on my env.

Besides, I encounter another case failure TestDevicePluginReRegistrationProbeMode. This is not related to this fix.
```
--- FAIL: TestDevicePluginReRegistrationProbeMode (2.05s)
    manager_test.go:225:
        	Error Trace:	manager_test.go:225
        	Error:      	Not equal:
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestDevicePluginReRegistrationProbeMode
        	Messages:   	Devices of previous registered should be removed
```
https://storage.googleapis.com/k8s-gubernator/triage/index.html?text=TestDevicePluginReRegistrationProbeMode#87728ff47d48e4a4cc99

**Comment by ehashman** [comment]
/priority awaiting-more-evidence
/triage accepted

Obviously we don't wanna panic so I think we should fix this if possible, I am just waiting on impact.

**Comment by klueska** [comment]
I still need to understand why this is panicking before we move forward to merge this. The current implementation of Allocate() assumes it is called serially (so there should be no opportunity a concurrent map access to occur, let alone panic). 

If Allocate() is indeed being called concurrently, then we need to redesign / rewrite the logic here completely (not just add a few locks).

Are these panics happening in a live deployment or in the unit tests? Maybe the unit tests are running a non-realistic setup?

**Comment by pacoxu** [comment]
Hold this as a draft since https://github.com/kubernetes/kubernetes/issues/102517#issuecomment-861108119. The issue creator cannot reproduce it. 
@klueska

**Comment by klueska** [comment]
@pacoxu I see you reopened this issue? Was it reproduced somewhere?

**Comment by pacoxu** [comment]
> @pacoxu I see you reopened this issue? Was it reproduced somewhere?

Not yet.
Sorry for the disturb. I reopened it because I mistakenly removed the branch and it is closed automaticly.

**Comment by pacoxu** [comment]
See #103838

**Comment by fromanirh** [comment]
we have evidence a panic triggers but I concur with https://github.com/kubernetes/kubernetes/pull/102525#issuecomment-854988681 - some additional investigation is warranted because we need to check some basic assumptions still holds true, otherwise just adding locks won't really solve.

**Comment by pacoxu** [comment]
ping @klueska as this PR may solve #103838 as well.

**Comment by thomas-riccardi** [comment]
I don't know about the status of this PR, but the issue it's trying to fix (#103838) is still occurring to us regularly in production.

/remove-lifecycle rotten

**Comment by klueska** [comment]
I will look into this next week and any fixes needed will be backported appropriately. Sorry for the delay.

**Comment by klueska** [comment]
As I hinted at in https://github.com/kubernetes/kubernetes/pull/102525#issuecomment-852947702, the `Allocate()` code (and its use of `reusableDevices`) has been designed assuming that it is **only** ever called by the `TopologyManager` and that the following pseudocode is run *atomically* across all containers in a pod before moving on to the next pod:

```
for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
    allHints = {}
    for _, provider := range hintProviders {
        hints := provider.GetTopologyHints(pod, container)
        allHints = append(allHints, hints)
    }

    bestHint := policy.Merge(allHints)

    for _, provider := range s.hintProviders {
        provider.Allocate(pod, container, bestHint)
    }
}
```

It is necessary to run this atomically to ensure that the hints produced in the first "provider" loop are still valid when doing the allocations within the second "provider" loop.

In general, this atomicity condition holds true because pod admission is serialized by the kubelet, so only one pod will ever be inside this loop at a given time.

This is what led me to state previously that:
> ... I’m still concerned about the need for this at all. If calls to `Allocate()` are not serialized, then the logic here has bigger issues than simply triggering a fault on a concurrent map access.

Knowing this "atomicity" requirement, I assumed that the only possible call to `Allocate()` would be in the `TopologyManager` code where the loop mentioned above is implemented. If that were true, then it should be "impossible" to encounter the concurrent map access this issue describes because we know that calls to this loop are serialized across pod admissions.

However .... digging into the code a bit, this is not actually true ...

There is one other place that `Allocate()` is called and that's in the body of the `GetDeviceRunContainerOptions()` call under one very specific condition.

Quoting from the body of `GetDeviceRunContainerOptions()`:
```
        // This is a device plugin resource yet we don't have cached
        // resource state. This is likely due to a race during node
        // restart. We re-issue allocate request to cover this race.
```

And this is ultimately triggered along the following call-graph from a call to `UpdatePod()` by the `kubelet`:
```
devicemanager.(*ManagerImpl).Allocate()
devicemanager.(*ManagerImpl).GetDeviceRunContainerOptions()
cm.(*containerManagerImpl).GetResources()
kubelet.(*Kubelet).GenerateRunContainerOptions()
kuberuntime.(*kubeGenericRuntimeManager).generateContainerConfig()
kuberuntime.(*kubeGenericRuntimeManager).startContainer()
kuberuntime.(*kubeGenericRuntimeManager).SyncPod.func1()
kuberuntime.(*kubeGenericRuntimeManager).SyncPod()
kubelet.(*Kubelet).syncPod()
kubelet.(*Kubelet).syncPod-fm()
kubelet.(*podWorkers).managePodLoop.func1()
kubelet.(*podWorkers).managePodLoop()
kubelet.(*podWorkers).UpdatePod.func1()
```

I will need to investigate further to figure out why this race occurs during the call to `UpdatePod()` and how we can avoid it to restore our atomicity guarantees. 

But one thing I know for sure -- just adding a simple mutex around our access to `reusableDevices` is *not* the proper solution for this.

**Comment by klueska** [comment]
@thomas-riccardi @tongchao199 @SidneyShen
May I ask which version of kubernetes you are running that exhibits this issue?

I ask because (based on my analysis above) I can't see a situation where `UpdatePod()` would be called on a not-yet-admitted pod in order to trigger the race.

However, in https://github.com/kubernetes/kubernetes/pull/103979 we fixed a bug that would cause a pod's `devicemanager` state to be prematurely cleared before it was supposed to be. If `UpdatePod()` was called on one of these pods that had its state cleared before it was properly terminated, then I could see how we might run into this weird edge case of concurrently calling `Allocate()` (when we actually shouldn't be at all).

From what I can tell the code in `GetDeviceRunContainerOptions()` that actually calls `Allocate()` was added at a time before we actually did allocation at pod-admit time. We used to do device allocation only upon container start, which meant that `UpdatePod()` might end up being called before the "normal" path for allocate was called, triggering it to do it "on-demand" in the `GetDeviceRunContainerOptions()` call and then skipping it in the "normal" path.

Now that we've moved to a model where allocation is now done before pod admission, I believe this call to `Allocate()` is actually unnecessary because it should be impossible to *not* have resources associated with a container at the time an `UpdatePod()` is called. Except of course when you have a bug and those resources are prematurely removed from the container before it has properly terminated.

As such, I'd be curious to know if running a version of kubernetes that includes https://github.com/kubernetes/kubernetes/pull/103979 already fixes the issue, and if so, then we should actually look into removing the code that calls allocate here to prevent future confusion.

**Comment by thomas-riccardi** [comment]
@klueska we are seeing this issue on `v1.19.14-gke.1900`; if I understand things correctly, #103979 landed in 1.23 and was backported to 1.22, so the answer would be: no idea if #103979 fixes the issue (unless of course if GKE backports such large PR on their fork...)

**Comment by klueska** [comment]
@pradvenkat do you know the process for backporting fixes to the versions of K8s running on GKE? If we backported #103979 to the next patch version of 1.19 would that get picked up more quickly than if we backported it to 1.20 or 1.21?

**Comment by dims** [comment]
There is a cncf-cla:no label on this PR. Can you please see the bot comments above to fix the CLA? This PR cannot progress unless the bot is happy with the CLA situation

**Comment by pacoxu** [comment]
/remove-lifecycle stale
/check-cla
/esaycla

#108583 fixed https://github.com/kubernetes/kubernetes/issues/103838, and also add some test cases

**Pull request closed without merging**